src/node/historical_queries.h

```diff
@@ -50,128 +58,296 @@ namespace ccf::historical
       {}
     };
 
-    struct Request
-    {
-      RequestStage current_stage = RequestStage::Fetching;
-      crypto::Sha256Hash entry_hash = {};
-      StorePtr store = nullptr;
-
-      // Only set when recovering ledger secrets
-      std::unique_ptr<LedgerSecretRecoveryInfo> ledger_secret_recovery_info =
-        nullptr;
-    };
-
-    // These constitute a simple LRU, where only user queries will refresh an
-    // entry's priority
-    static constexpr size_t MAX_ACTIVE_REQUESTS = 10;
-
-    using Requests = std::map<consensus::Index, Request>;
-    Requests requests;
-    std::list<consensus::Index> recent_requests;
-
-    // To trust an index, we currently need to fetch a sequence of entries
-    // around it - these aren't user requests, so we don't store them, but we do
-    // need to distinguish things-we-asked-for from junk-from-the-host
-    std::set<consensus::Index> pending_fetches;
-
-    ccf::VersionedLedgerSecret get_first_known_ledger_secret()
+    ccf::VersionedLedgerSecret get_earliest_known_ledger_secret()
     {
       if (historical_ledger_secrets->is_empty())
       {
-        return network.ledger_secrets->get_first();
+        return source_ledger_secrets->get_first();
       }
 
-      auto tx = network.tables->create_read_only_tx();
+      auto tx = source_store.create_read_only_tx();
       CCF_ASSERT_FMT(
         historical_ledger_secrets->get_latest(tx).first <
-          network.ledger_secrets->get_first().first,
+          source_ledger_secrets->get_first().first,
         "Historical ledger secrets are not older than main ledger secrets");
 
       return historical_ledger_secrets->get_first();
     }
 
-    void request_entry_at(consensus::Index idx)
+    struct StoreDetails
+    {
+      RequestStage current_stage = RequestStage::Fetching;
+      crypto::Sha256Hash entry_digest = {};
+      StorePtr store = nullptr;
+      bool is_signature = false;
+    };
+    using StoreDetailsPtr = std::shared_ptr<StoreDetails>;
+
+    struct Request
     {
-      // To avoid duplicates, remove index if it was already requested
-      recent_requests.remove(idx);
+      consensus::Index first_requested_idx = 0;
+      consensus::Index last_requested_idx = 0;
+      std::vector<StoreDetailsPtr> requested_stores;
+      std::chrono::milliseconds time_to_expiry;
 
-      // Add request to front of list, most recently requested
-      recent_requests.emplace_front(idx);
+      // An entry from outside the requested range containing the next signature
+      // may be needed to trust this range. It is stored here, distinct from
+      // user-requested stores.
+      std::optional<std::pair<consensus::Index, StoreDetailsPtr>>
+        supporting_signature;
```

Comment:
`evidence` would be preferable, for CFT this is a single signature, but this is really "consensus evidence" in general.